lib/saluki-components/src/transforms/metric_router/mod.rs

```diff
@@ -0,0 +1,250 @@
+use serde::Deserialize;
+use tokio::select;
+use tracing::{debug, error};
+
+/// Metric router transform configuration.
+#[derive(Deserialize, Clone)]
+pub struct MetricRouterConfiguration {
+    /// List of metric names to match against.
+    pub metric_names: Vec<String>,
+}
+
+/// Routes incoming events based on exact metric name matching to one of two
+/// named outputs: "matched" for metrics with names that exactly match the
+/// configured list, and "unmatched" for all other events. This transform
+/// demonstrates the use of named outputs without a default output.
+#[derive(Debug)]
+pub struct MetricRouter {
+    metric_names: HashSet<String>,
+}
+
+impl MetricRouter {
+    fn new(config: MetricRouterConfiguration) -> Result<Self, GenericError> {
+        for name in &config.metric_names {
+            if name.trim().is_empty() {
+                return Err(generic_error!("Metric names cannot be empty or whitespace-only"));
+            }
+        }
+
+        Ok(Self {
+            metric_names: config.metric_names.into_iter().collect(),
+        })
+    }
+
+    /// Evaluates whether the event should be routed to the matched output.
+    ///
+    /// Returns true if the event is a metric with a name that exactly matches
+    /// one of the configured metric names, false otherwise.
+    fn should_route_to_matched(&self, event: &Event) -> bool {
+        match event {
+            Event::Metric(metric) => {
+                // Check if the metric name exactly matches any of our configured names
+                let metric_name = metric.context().name();
+                self.metric_names.contains(metric_name.as_ref())
+            }
+            _ => {
+                // Non-metric events are always routed to unmatched
+                false
+            }
+        }
+    }
+
+    /// Processes a batch of events by routing them to matched or unmatched outputs.
+    async fn process_event_batch(
+        &self, events: EventsBuffer, context: &mut TransformContext,
+    ) -> Result<(), GenericError> {
+        let mut matched_dispatcher = context.dispatcher().buffered_named("matched")?;
+        let mut unmatched_dispatcher = context.dispatcher().buffered_named("unmatched")?;
+
+        let mut matched_count = 0;
+        let mut unmatched_count = 0;
+
+        // Route each event to the appropriate output
+        for event in events.into_iter() {
+            if self.should_route_to_matched(&event) {
+                matched_count += 1;
+                matched_dispatcher.push(event).await?;
+            } else {
+                unmatched_count += 1;
+                unmatched_dispatcher.push(event).await?;
+            }
+        }
+
+        matched_dispatcher.flush().await?;
+        unmatched_dispatcher.flush().await?;
+
+        debug!(
+            matched_events = matched_count,
+            unmatched_events = unmatched_count,
+            "Successfully processed event batch"
+        );
+
+        Ok(())
```

Comment:
I think could optimize this slightly for the cases where no metrics are being matched: use `FixedSizeEventBuffer::extract` to avoid allocating a brand new event buffer for the unmatched metrics, and then just pass the original/remaining event buffer via `context.dispatcher().dispatch_named(...)`.

In that case, we only pay the penalty of iterating over the metrics... but the iterator won't return anything, and so the buffered dispatcher will never allocate anything, and we just end up using the input event buffer as-is.